README.md
# Critical Fumble
Do you find yourself forgetting to roll on the critical hit table? The critical fumble table?
Do you find yourself forgetting to roll that pesky loot table after each monster dies?

Then you're like me and this mod is what you need!
Critical-fumble will automatically roll on the critical hit tables in the event of a critical hit! Each hit rolls a special table depending on the type of damage.

Additionally, every time a *targeted* monster is felled by your players, the CR table is rolled! 
To further your convience, the coins dropped can be automatically distributed to your players. This option can be ignored though! 

## Features:
See the [examples](##Examples) for video demonstrations!
The following features are offered by Critical Fumble. Any feature marked with a **(T/F)** can be disabled or enabled.
1. Critical hits on 20s for any of the values in the [Critical Tables](##Tables) **(T/F)** 
2. Critical fumbles on 1s **(T/F)**
3. Audio based off the damage type on critical hits **(T/F)**
4. LootTable rolling on target creature death **(T/F)**
5. LootTable distribution of wealth to active PCs **(T/F)**
6. Private rolls on both Critical Fumble and Critical Fumble Loot **(T/F)**

## Incompatibilities

Due the requirement of knowing the damage type, I overloaded Dice5e.roll(). If any module overrides this method as well, they will be incompatible. 

## Installation:

To install Critical-Fumble, do the following:

1. [Download the zip file](https://github.com/JacobMcAuley/critical-fumble/archive/master.zip)
2. Extract the folder to your 'public/modules' in Foundry VTT.
3. Reload your application.

Auto-Installation Instructions:

1. Copy this link: https://github.com/JacobMcAuley/critical-fumble/raw/master/module.json
2. Open FoundryVTT and navigate to Add-On Modules
3. Select "Install Module" and paste link into textbox.
4. Click "Install" and watch as the mod is installed
5. Inside of your FVTT game world, navigate the the gears in the top right corner.
6. Select "Manage Modules" and enable Critical Fumble

## Usage

To use Critical-Fumble make sure you've installed the mod using the instructions [here](##Installation).
Critical fumble should be loaded and ready to go upon enabling. To verify, simply check to see if the RollTables were successfully imported. (**NOTE:** You may find that the values display an incorrect number of entries in the table. This is okay and can be disregarded).

**Using Criticals**
To use Critical hits, wait until a player rolls a 1 or the critical range (Player dependent) and the table will be automatically rolled for you!
For spells, be sure that a ranged spell attack is set up, otherwise no table will be rolled.

**Using the LootTable**
Creatures (NPCS) that have a CR level and are targeted using the targeting system will result in a roll on their corresponding CR table once they hit 0hp or lower.

## Tables
The follow tables are included within the module and imported upon start:

**Criticals**
1. Bludgeon
2. Piercing
3. Slashing
4. Acid
5. Cold
6. Fire
7. Force
8. Lightning
9. Necrotic
10. Poison
11. Psychic
12. Radiant
13. Thunder

**Special Tables**
1. Major Injuries
2. Minor Injuries
3. Insanity

**Loot**
1. CR 0-4
2. CR 5-10
3. CR 11-16
4. CR 17+

**Fumbles**
1. Fumble

Currently, there is no table for fumbles on spells. All fumbles share the same Critical-Fumble Fumble table. If anyone is interested in creating a relatively balanced table and would like to incorporate fumbles on spells, feel free to contact me through Discord found through the Feedback section below.


### Table modifications?

If you would like to adjust the table, you should be able to in most instances! Due to recent structure changes in the program. You'll need to adjust any changes in the .db files in the module directory. Look for any of the optional flags for any of the loot based rolls.

Otherwise, you can modify the values in the tables as desired via FVTT.

## Attributions

Super special thanks to [Kakaroto](https://www.patreon.com/kakaroto/) for his code review!
Sound effects obtained from [Zapsplat](https://www.zapsplat.com)  
SVGs obtained from [FontAwesome](https://fontawesome.com). You can find the license by clicking [here](https://fontawesome.com/license)  
Icon-Sword made by [freepik](https://www.flaticon.com/authors/freepik) from [flaticon](https://www.flaticon.com/)  
Icon-dagger made by [srip](https://www.flaticon.com/authors/srip) from [flaticon](https://www.flaticon.com/)  
Icon-psychic made by [Smashicons](https://www.flaticon.com/authors/Smashicons) from [flaticon](https://www.flaticon.com/)  
Critical tables made by [Benjamin Huffman](https://sterlingvermin.files.wordpress.com/2016/09/critical-hits-revisited.pdf)  
Map used in video made by [Cze and Peku](https://www.patreon.com/czepeku/posts)

## Feedback

This module is still in the testing phase and is likely to have unexpect bugs or weird interactions. Please remember to back up your systems before trying new modules. If any gamebreaking or unexpected behavior occurs, please contact me on Discord through PM or the FVTT (preferred) with the name JacobMcAuley#3461. Thank you for trying out Critical-Fumble!


# Donations
If you liked this project and want to support future ones, consider donating! 
This module and all previous and future modules, as always, will be free to use and download. Any donations are warmly received and appreciated!
[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=723SW7WMD8YR6&item_name=Thank+you+for+your+tip%21&currency_code=USD&source=url)


## Examples

![Crit_Fire](examples/Crit.gif)
![Crit_Loot](examples/Loot.gif)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.6.3] - 2020-01-06
### Special Thanks
- This patch was inspired/review by Kakaroto from FVTT (https://github.com/kakaroto), a brilliant developer and a wonderful person in the FVTT community. Consider checking out his github or even his Patreon (https://www.patreon.com/kakaroto/). 

### Added
- Private roll options for GMs who would like to keep the table roll a secret.
- Tables are now placed into folders for added organization.
- A "coin pile" has been added to persist over game sessions. If an uneven amount of coins are distributed IE: Three PCs and 13SP, the remaining coin will sit in a 'pot' until the first possible moment in which all players can receive the same amount.
    EX: a monster is slain for 13SP. Given three players, each receives 4SP and one is stored in the SP coin pile. Later, another monster is slain for 15SP. Each player receives 4SP. The 2SP remaining is distributed to the SP coin pile. It is then determined automatically that 3SP {1SP from earlier and 2SP now}can be evenly distributed across the players, thus instead of 4SP for this encounter, 5SP is awarded instead.

### Changed
- Adjusted tables to now be included as compendium packs, opposed to programmatically generated.
- Critical hits are now based on Actor critical hit range and not just 20s.
- Reorganized the JS to now be more readable and improved cleanliness.
- Adjusted audio play to use FVTT's build in play.
- Changed how coins are given out. See "Coin Pile" in added.

### Fixed
- Fixed bug where the table would be rolled for each user in the game
- Fixed bug where other rolls(Observer, Assistant GM) were not accounted for in wealth distribution and thus less coins were given.
- Fixed bug where private rolls could be unintentionally blocked.

## [0.5.0] - 2019-12-28
### Added
- Critical Tables for both fumbles and successes
- Loot Drop Tables upon creature kill
- Uses targeting to determine creature drop tables.
- Intercepts attack rolls, and automatically rolls on table if either a 1 or 20 was rolled.
- Settings added to disable select features: Loot Tables, Critical Success, Critical Failure.

src/main.js
///RollTable.collection.entities.find(function(table){ return table.data.name == "Critical-Fumble Critical"})

class CriticalFumble{
    constructor(){
        this.tables = 0 //this.findTable();
        this.init()
    }

    async findTable(){

        let critical = RollTable.collection.entities.find(function(table){ return table.data.name == "Critical-Fumble Critical"});
        let fumble = 
        Object.keys(CRITICAL_FUMBLE_DESC).forEach(function(tableName){
            let tempTable = tableName.charAt(0).toUpperCase()
            if(!RollTable.collection.entities.find(function(table){ return table.data.name == `Critical-Fumble ${tempTable}`})){ // Doesn't return a value
                await this._generateTable(`Critical-Fumble ${tempTable}`, CRITICAL_FUMBLE_DESC[tableName]);
            }
        })        
        return [critical, fumble];
    }

    async _generateTable(tableName, data){
        let table = await RollTable.create({name: tableName, type: "base", folder: null, types: "base", formula: "1d100"}, {displaySheet: false});

        await table.createTableResult({ collection: undefined,drawn: false, range: [NUM1, NUM2],type: 0,weight: 1,text: itemDescriptions[i-1]})
    }

    init(){
        Hooks.on('createChatMessage', (app, html, data) => {
            try{
                this.validateData(app);
            }
            catch(e){
                console.log(`Nope: ${e}`);
            }
        })
    }

    validateData(app){
        console.log(app);
        if(!this._validString(app.data.flavor))
            return;
        let roll = JSON.parse(app.data.roll)
        let index = this._validThreshold()
        console.log("Success!")
    }

    _validString(string){
        if(!string)
            return false;
        if(string.includes("Attack"))
            return true;

    }

    _validThreshold(value){
        return true;
    }

    async roll(index){
        let [roll, result] = this.table[index].roll()
        if (result !== null ) {
            await this.table[index].draw({roll, result});
        }
    }
}

Hooks.on("ready", () => {
    CriticalFumble = new CriticalFumble();
    /*
    let [roll, result] = RollTable.collection.entities[0].roll()
    if ( result !== null ) {
        await RollTable.collection.entities[0].draw({roll, result});
    }
    */
})



config/config.js
export const CRITICAL_FUMBLE_CONFIG = [
    {
        module : "critical-fumble",
        key: "criticalTable",
        settings : {
            name : "Enables critical table",
            hint : "Allow for the critical table to be rolled on Natural 20s",
            scope : "world",
            config : true,
            default : true,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "fumbleTable",
        settings : {
            name : "Enables fumble table",
            hint : "Allow for the fumbe table to be rolled on Natural 1s",
            scope : "world",
            config : true,
            default : true,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "sounds",
        settings : {
            name : "Enable Rolltable sounds",
            hint : "Allows for sounds to be played during table hits",
            scope : "world",
            config : true,
            default : true,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "lootTables",
        settings : {
            name : "Enable Loot tables",
            hint : "Allow for the Loot Table to be rolled on NPC death",
            scope : "world",
            config : true,
            default : true,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "coinDistribution",
        settings : {
            name : "Enable coin distribution",
            hint : "Enable the coin distribution prompt",
            scope : "world",
            config : true,
            default : true,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "privateHitTableRolls",
        settings : {
            name : "Enable private critical/fumble rolls",
            hint : "Enabling this will make all critical/fumbles rolls private",
            scope : "world",
            config : true,
            default : false,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "privateLootRolls",
        settings : {
            name : "Enable private loot rolls",
            hint : "Enabling this will make all loot rolls private",
            scope : "world",
            config : true,
            default : false,
            type : Boolean,
            choices : undefined 
        }
    },
    {
        module : "critical-fumble",
        key: "coinPileCP",
        settings : {
            name : "A coin pile for even distribution",
            hint : "Collects left over coins until they can be evenly distributed.",
            scope : "world",
            default : 0,
            type : Object,
            choices : undefined 
        }  
    },
    {
        module : "critical-fumble",
        key: "coinPileSP",
        settings : {
            name : "A coin pile for even distribution",
            hint : "Collects left over coins until they can be evenly distributed.",
            scope : "world",
            default : 0,
            type : Object,
            choices : undefined 
        }  
    },
    {
        module : "critical-fumble",
        key: "coinPileGP",
        settings : {
            name : "A coin pile for even distribution",
            hint : "Collects left over coins until they can be evenly distributed.",
            scope : "world",
            default : 0,
            type : Object,
            choices : undefined 
        }  
    },
    {
        module : "critical-fumble",
        key: "coinPilePP",
        settings : {
            name : "A coin pile for even distribution",
            hint : "Collects left over coins until they can be evenly distributed.",
            scope : "world",
            default : 0,
            type : Object,
            choices : undefined 
        }  
    },   
]

src/criticalFumble.js
import {CriticalFumbleBase} from "./criticalFumbleBase.js"

class CriticalFumble extends CriticalFumbleBase{
    constructor(){
        super();
        this.folderName = "Critical Fumble";
        this.color = "green";
        this.compendiumName = "criticalfumble";
        this.moduleScope = "critical-fumble";
        this.privateRollsSettingName = "privateHitTableRolls";
        this.init();
    }

    init(){
        super.init();
        Hooks.on('createChatMessage', this.handleAttackRoll.bind(this));
    }

    handleAttackRoll(messages, data, options, id){
        if(!this.isAttackRoll(getProperty(data, "flavor")))
            return;
        let parsedRollData = JSON.parse(getProperty(data, "roll"))
        if(!parsedRollData)
            return;
        let type = parsedRollData.parts[0].options.damageType;
        let criticalRange = parsedRollData.parts[0].options.critical;
        let roll = this._getRoll(parsedRollData.parts[0].rolls);
        if(this.debug)
            roll = this.debugRoll;
        let table = this._validThreshold(roll, this.capitalizeFirst(type), criticalRange)
        if(table)
            this.rollTable(table);
    }

    isAttackRoll(string){
        if(!string)
            return false;
        if(string.includes("Attack"))
            return true;
    }

    _getRoll(rolls){
        for(let i = 0; i < rolls.length; ++i){
            if(rolls[i].discarded != true)
                return rolls[i].roll;
        }
        return -1;
    }

    _validThreshold(roll, type, criticalRange){
        if(roll == 1 && game.settings.get('critical-fumble', 'fumbleTable'))
            return this.tables.find(tableName => tableName.data.name === "Critical-Fumble Fumble");
        else if(roll >= criticalRange && game.settings.get('critical-fumble', 'criticalTable'))
            return this.tables.find(tableName => tableName.data.name === `Critical-Fumble ${type}`);
        return undefined;      
    }

    async rollTable(table){
        let [roll, result] = table.roll();
        if (result !== null ) {
            await this._customDraw({roll, result}, table);
        }
        if(game.settings.get('critical-fumble', 'sounds')){
            AudioHelper.play({src: result.flags["critical_fumble_audio"]}); 
        }
    }

    capitalizeFirst(name){
        return name.charAt(0).toUpperCase() + name.slice(1);
    }

    get privateRollSetting(){
        return game.settings.get(this.moduleScope, this.privateRollsSettingName)
    }

}

Hooks.on("ready", () => {
    if(game.user.isGM)
        CriticalFumble = new CriticalFumble();
})



src/criticalFumbleBase.js
import { Dice5e } from "../../../systems/dnd5e/module/dice.js";
import {CRITICAL_FUMBLE_CONFIG} from "../config/config.js"

export class CriticalFumbleBase{
    constructor(){
        this.debug = false;
        this.debugRoll = 20;
        this.folderID;
        this.tables = [];
    }

    init(){
        this.ensureTableImport();
    }

    async ensureTableImport(){
        this.folderID = await this.handleFolder(); 
        this.ensureTables();
    }

    async handleFolder(){
        let folder = Folder.collection.entities.find(folder => {return folder.name === this.folderName});
        if(!folder)
            return (await Folder.create({name : this.folderName, type: "RollTable", color: this.color, parent:null})).id;
        else   
            return folder.id;
    }

    async ensureTables(){
        let criticalCompendium = game.packs.find(compendium => compendium.metadata.name === this.compendiumName);
        await criticalCompendium.getIndex();
        criticalCompendium.index.forEach(async tableEntry => {
            let resultingTable = RollTable.collection.entities.find(function(table){ return table.name == tableEntry.name})
            if(!resultingTable)
                resultingTable = await this._generateTable(criticalCompendium, tableEntry.id);
            this.tables.push(resultingTable);
        });
    }

    async _generateTable(pack, entryID){
        let tableData = (await pack.getEntity(entryID)).data;
        tableData.folder = this.folderID;
        await RollTable.create(tableData, { displaySheet: false});
    }

    async _customDraw({roll=null, result=null}={}, table){
        const speaker = "Gamemaster";
        if ( roll === result === null ) {
          [roll, result] = this.roll();
        }
        if ( result === null ) return;
    
        let rollOption = game.settings.get("core", "rollMode");
        if(this.privateRollSetting)
            rollOption = "gmroll";
            
        await roll.toMessage({
            speaker: speaker,
            flavor: `Draws a result from ${table.name}`,
            rollMode: rollOption
        });
        
        await this._displayChatResult(result, speaker, table);
        return result;
    }

    _displayChatResult(result, speaker, table) {

        // Basic chat message data
        const chatData = {
          user: game.user._id,
          type: CONST.CHAT_MESSAGE_TYPES.OTHER,
          speaker: speaker
        };
    
        // Toggle default roll mode
        let rollMode = game.settings.get("core", "rollMode");
        if(this.privateRollSetting)
            rollMode = "gmroll";
        if ( ["gmroll", "blindroll"].includes(rollMode) ) chatData["whisper"] = ChatMessage.getWhisperIDs("GM");
        if ( rollMode === "blindroll" ) chatData["blind"] = true;
    
        // Toggle the result format
        let text = result.text;
        if ( result.type === CONST.TABLE_RESULT_TYPES.ENTITY ) {
          text = `@${result.collection}[${result.text}]`;
        }
    
        // Render the template
        chatData["content"] = `
        <div class="table-result" data-table-id="${table._id}" data-result-id="${result.id}">
            <img class="result-image" src="${result.img || CONFIG.RollTable.resultIcon}"/>
            <h4 class="result-text">${text}</h4>
        </div>`;
    
        // Create the chat message
        return ChatMessage.create(chatData);
    }


    static rollOverload(){
        Dice5e.d20Roll = function({event, parts, data, template, title, speaker, flavor, advantage=true, situational=true,
            fastForward=true, critical=20, fumble=1, onClose, dialogOptions, }) {
            flavor = flavor || title;

            // Inner roll function
            let rollMode = game.settings.get("core", "rollMode");
            let roll = (parts, adv) => {
            if (adv === 1) {
            parts[0] = ["2d20kh"];
            flavor = `${title} (Advantage)`;
            }
            else if (adv === -1) {
            parts[0] = ["2d20kl"];
            flavor = `${title} (Disadvantage)`;
            }

            // Don't include situational bonus unless it is defined
            if (!data.bonus && parts.indexOf("@bonus") !== -1) parts.pop();

            // Execute the roll
            let roll = new Roll(parts.join(" + "), data).roll();

            // Flag critical thresholds
            let d20 = roll.parts[0];
            d20.options.critical = critical;
            d20.options.fumble = fumble;
            try{
                let damageType = data.item.damage.parts[0][data.item.damage.parts[0].length - 1];
                d20.options.damageType = damageType;
            }
            catch{}

            
          // This is the overload
            // Convert the roll to a chat message
            roll.toMessage({
                speaker: speaker,
                flavor: flavor,
                rollMode: rollMode,
                });
            };

            // Modify the roll and handle fast-forwarding
            parts = ["1d20"].concat(parts);
            if ( event.shiftKey ) return roll(parts, 0);
            else if ( event.altKey ) return roll(parts, 1);
            else if ( event.ctrlKey || event.metaKey ) return roll(parts, -1);
            else parts = parts.concat(["@bonus"]);

            // Render modal dialog
            template = template || "systems/dnd5e/templates/chat/roll-dialog.html";
            let dialogData = {
            formula: parts.join(" + "),
            data: data,
            rollMode: rollMode,
            rollModes: CONFIG.rollModes
            };
            let adv = 0;
            renderTemplate(template, dialogData).then(dlg => {
            new Dialog({
                title: title,
                content: dlg,
                buttons: {
                advantage: {
                    label: "Advantage",
                    callback: () => adv = 1
                },
                normal: {
                    label: "Normal",
                },
                disadvantage: {
                    label: "Disadvantage",
                    callback: () => adv = -1
                }
                },
                default: "normal",
                close: html => {
                if ( onClose ) onClose(html, parts, data);
                rollMode = html.find('[name="rollMode"]').val();
                data['bonus'] = html.find('[name="bonus"]').val();
                roll(parts, adv);
                }
            }, dialogOptions).render(true);
            });
        }
    }
}

Hooks.on('init', () => {
    CRITICAL_FUMBLE_CONFIG.forEach(function(setting){
        game.settings.register(setting.module, setting.key, setting.settings);
    });
    CriticalFumbleBase.rollOverload();
})

src/criticalLoot.js

import {CriticalFumbleBase} from "./criticalFumbleBase.js"

class CriticalLoot extends CriticalFumbleBase{
    constructor(){
        super();
        this.folderName = "Critical Fumble Loot";
        this.color = "red";
        this.compendiumName = "criticalfumbleloot";
        this.moduleScope = "critical-fumble";
        this.privateRollsSettingName = "privateLootRolls";
        this.init();
    }

    init(){
        super.init();
        Hooks.on("updateToken", this._onTokenUpdate.bind(this))
    }

    _onTokenUpdate(sceneData, sceneID, tokenData){
        let actor = game.actors.tokens[tokenData._id];
        if(!game.settings.get(this.moduleScope, 'lootTables'))
            return
        if(!actor || actor.isPC)
            return;
        if(!this.isValidTarget(actor))
            return
        if(!this.isTokenKilled(tokenData))
            return
        let table = this.getTableByCR(actor)
        if(table && game.user.isGM)
            this.rollTable(table);
    }

    
    isValidTarget(actor){
        let targets = []
        game.users.entities.forEach(user => { user.targets.forEach(target => {targets.push(target)}) })
        for(let i = 0; i < targets.length; ++i){
            if(targets[i].actor.id == actor.id)
                return true;
        }
        return false;
    }

    getTableByCR(actor){
        let cr = getProperty(actor, "data.data.details.cr");
        if(cr >= 0 && cr <= 4)
            return this.tables.find(table => table.name === `Critical-Fumble CR_0_4`);
        else if(cr >= 5 && cr <= 10)
            return this.tables.find(table => table.name === `Critical-Fumble CR_5_10`);
        else if(cr >= 11 && cr <= 16)
            return this.tables.find(table => table.name === `Critical-Fumble CR_0_4`);
        else if(cr >= 17)
            return this.tables.find(table => table.name === `Critical-Fumble CR_17`);
        throw "Critical Fumble : Unexpected CR Value" // Helps diagnose possible errors.
    }

    isTokenKilled(tokenData){
        let hp = getProperty(tokenData, "actorData.data.attributes.hp.value");
        return hp !== undefined && hp <= 0;
    }

    async rollTable(table){
        let [roll, result] = table.roll()
        if(!result)
            throw "Critical Fumble : Undefined result error" // Helps diagnose possible errors.
        result.img = "modules/critical-fumble/icons/coins.svg";
        await this._customDraw({roll, result}, table);
        let parsedRolls = getProperty(result.flags, "critical_fumble_actions") || []; 
        this.handleRolls(parsedRolls); 
    }

    handleRolls(rolls){
        let coins = []
        
        for(let i = 0; i < rolls.length; ++i){
            let total = this._roll(rolls[i].roll, rolls[i].description, rolls[i].multiplier); 
            coins.push({coinAmount: total, coinType: rolls[i].description});
        }
        if(game.settings.get('critical-fumble', 'coinDistribution'))
            this._coinDistributePrompt(coins)
    }

    _roll(newRoll, coin, multiplier){
        let roll = new Roll(`${newRoll} * ${multiplier}`).roll();
        let rollMode = "roll";
        if(this.privateRollSetting)
            rollMode = "gmroll";
        let speaker = "Critical Fumble";
        let flavor = `Rolls for ${coin}`
        let sound = undefined;
        if(game.settings.get('critical-fumble', 'sounds'))
            sound = "modules/critical-fumble/sounds/money_drop.wav"

        roll.toMessage({
          speaker: speaker,
          flavor: flavor,
          rollMode: rollMode,
          sound: sound
        });

        return roll.total;
    };
    
    _coinDistributePrompt(coins){
        let amountDisplay = this._coinToString(coins);
        new Dialog({
            title: "Distribute Coins?",
            content: `<p>Would you like to distribute ${amountDisplay} to all active players?</p>`,
            buttons: {
             yes: {
              icon: '<i class="fas fa-check"></i>',
              label: "Yes, distribute the coins.",
              callback: () => this._coinDistribute(coins)
             },
             no: {
              icon: '<i class="fas fa-times"></i>',
              label: "No, I will handle it later.",
              callback: () => {}
             }
            },
            default: "yes",
            close: () => {}
        }).render(true);
    }

    _coinToString(coins){
        let string = "";
        for(let i = 0; i < coins.length; ++i){
            if(i == coins.length - 1){
                string += `${coins[i].coinAmount}${coins[i].coinType}.`
            }
            else{
                string += `${coins[i].coinAmount}${coins[i].coinType} + `
            }
        }
        return string;
    }

    async _coinDistribute(coins){
        let users = game.users.entities;
        let totalPlayers = game.users.entities.reduce((acc, user) => {
            if(!user.isGM && user.character)
                acc += 1; 
            return acc;
        }, 0)

        for(let i = 0; i < coins.length; ++i){
            let coinsToDistribute = this._coinCalculator(coins[i].coinAmount, totalPlayers, coins[i].coinType); 
            for(let j = 0; j < users.length; ++j){
                if(users[j].character != undefined && users[j].isGM == false){
                    let currency = duplicate(users[j].character.data.data.currency);
                    currency[coins[i].coinType.toLowerCase()] += coinsToDistribute;
                    await users[j].character.update({"data.currency" : currency});
                }
            }
        }
    }

    _coinCalculator(coinAmount, totalPlayers, coinType){
        let coinRemainderKey = `coinPile${coinType}`;
        let coinPile = game.settings.get(this.moduleScope, coinRemainderKey) || 0;
        let coinDistrubution = Math.floor(coinAmount / totalPlayers);
        coinPile += coinAmount % totalPlayers;
        if(coinPile != 0 && coinPile >= totalPlayers){
            let coinsAddedPerPlayer = Math.floor(coinPile / totalPlayers);
            coinDistrubution += coinsAddedPerPlayer
            game.settings.set(this.moduleScope, coinRemainderKey, coinPile - (coinsAddedPerPlayer * totalPlayers))
        }
        else
            game.settings.set(this.moduleScope, coinRemainderKey, coinPile)
        return coinDistrubution;
    }
    
    get privateRollSetting(){
        return game.settings.get(this.moduleScope, this.privateRollsSettingName)
    }

}

Hooks.on("ready", () => {
    if(game.user.isGM)
        CriticalLoot = new CriticalLoot();
})
